File: src/lib/dns/character-string.ts
/**
 * Helpers for DNS `<character-string>` RDATA (RFC 1035 §3.3, §5.1).
 *
 * Record types whose RDATA is built from character-strings (TXT, SPF, the
 * TXT-shaped helpers such as DKIM/DMARC, HINFO, NAPTR text fields and the CAA
 * value field) are written in *presentation format*: the payload may be bare,
 * wrapped in double quotes, or expressed as several adjacent quoted strings
 * that concatenate into one logical value.
 *
 * This module is the single place that understands that syntax:
 *
 * - {@link parseCharacterStrings} tolerantly decodes any of those shapes,
 *   repairing unmatched quotes instead of rejecting the input.
 * - {@link unquoteCharacterString} collapses the result into the logical value
 *   semantic parsers (SPF mechanisms, DMARC/DKIM tags, …) should look at.
 * - {@link normalizeCharacterString} serializes a value back into a thoroughly
 *   valid quoted form, escaping what must be escaped and splitting payloads
 *   longer than 255 *bytes* into adjacent quoted strings.
 */

/** Maximum size, in bytes, of a single DNS `<character-string>`. */
export const CHARACTER_STRING_MAX_BYTES = 255;

const DECIMAL_ESCAPE = /^[0-9]{3}$/u;

const isControlCodePoint = (codePoint: number) =>
  codePoint <= 0x1f || (codePoint >= 0x7f && codePoint <= 0x9f);

const utf8Size = (codePoint: number) => {
  if (codePoint <= 0x7f) return 1;
  if (codePoint <= 0x7ff) return 2;
  if (codePoint <= 0xffff) return 3;
  return 4;
};

/**
 * Size of `value` in UTF-8 bytes, which is what the 255 byte
 * `<character-string>` limit is measured in.
 */
export function characterStringByteLength(value: string | undefined): number {
  let size = 0;
  for (const character of value ?? "") {
    size += utf8Size(character.codePointAt(0) ?? 0);
  }
  return size;
}

/** Decode a single `\X` / `\DDD` escape starting at `index`. */
function decodeEscapeAt(input: string, index: number): [string, number] {
  const decimal = input.slice(index + 1, index + 4);
  if (DECIMAL_ESCAPE.test(decimal)) {
    const octet = Number.parseInt(decimal, 10);
    if (octet <= 255) return [String.fromCharCode(octet), index + 4];
  }
  const next = input[index + 1];
  if (next === undefined) return ["\\", index + 1];
  return [next, index + 2];
}

/** Decode every escape sequence in `value`. */
function decodeEscapes(value: string): string {
  if (!value.includes("\\")) return value;
  let decoded = "";
  let index = 0;
  while (index < value.length) {
    const character = value[index] ?? "";
    if (character === "\\") {
      const [text, next] = decodeEscapeAt(value, index);
      decoded += text;
      index = next;
      continue;
    }
    decoded += character;
    index++;
  }
  return decoded;
}

function scanQuotes(value: string): { count: number; lastIndex: number } {
  let count = 0;
  let lastIndex = -1;
  for (let index = 0; index < value.length; index++) {
    const character = value[index];
    if (character === "\\") {
      index++;
      continue;
    }
    if (character === '"') {
      count++;
      lastIndex = index;
    }
  }
  return { count, lastIndex };
}

/**
 * True when `raw` contains an opening quote without its closing counterpart
 * (or the other way around). Such input is repaired rather than rejected.
 */
export function hasUnbalancedQuotes(raw: string | undefined): boolean {
  const trimmed = (raw ?? "").trim();
  return scanQuotes(trimmed).count % 2 === 1;
}

/**
 * True when `raw` uses the quoted presentation form, including the damaged
 * shapes with only a leading or only a trailing quote.
 */
export function isQuotedForm(raw: string | undefined): boolean {
  const trimmed = (raw ?? "").trim();
  if (!trimmed) return false;
  if (trimmed.startsWith('"')) return true;
  const { count, lastIndex } = scanQuotes(trimmed);
  return count % 2 === 1 && lastIndex === trimmed.length - 1;
}

/** Read one quoted run starting just after its opening quote. */
function scanQuotedRun(
  input: string,
  start: number,
): { value: string; end: number; closed: boolean } {
  let value = "";
  let index = start;
  while (index < input.length) {
    const character = input[index] ?? "";
    if (character === "\\") {
      const [text, next] = decodeEscapeAt(input, index);
      value += text;
      index = next;
      continue;
    }
    if (character === '"') return { value, end: index + 1, closed: true };
    value += character;
    index++;
  }
  // Unmatched opening quote: repair by closing the string at end of input.
  return { value, end: index, closed: false };
}

/** Read one bare (unquoted) run up to the next unescaped quote. */
function scanBareRun(
  input: string,
  start: number,
  stopOnWhitespace: boolean,
): { value: string; end: number } {
  let raw = "";
  let index = start;
  while (index < input.length) {
    const character = input[index] ?? "";
    if (character === '"') break;
    if (stopOnWhitespace && /\s/u.test(character)) break;
    if (character === "\\") {
      raw += input.slice(index, index + 2);
      index += 2;
      continue;
    }
    raw += character;
    index++;
  }
  return { value: decodeEscapes(raw).trim(), end: index };
}

/**
 * Parse presentation-format content into its logical character-strings.
 *
 * Accepted shapes, all decoded to the same logical value:
 * - bare: `v=spf1 include:_spf.example.com ~all` → one string, kept verbatim
 * - quoted: `"v=spf1 include:_spf.example.com ~all"`
 * - adjacent strings: `"part one" "part two"` → two strings
 * - unmatched leading or trailing quote → repaired
 *
 * Escaped quotes (`\"`), escaped backslashes (`\\`) and decimal escapes
 * (`\010`) inside a quoted run are decoded, never treated as delimiters.
 * Whitespace inside quotes is preserved exactly.
 */
export function parseCharacterStrings(raw: string | undefined): string[] {
  const trimmed = (raw ?? "").trim();
  if (!trimmed) return [];

  if (!trimmed.startsWith('"')) {
    // Bare content is a single character-string: whitespace does not split it.
    if (isQuotedForm(trimmed)) {
      // Unmatched trailing quote: repair by treating it as a quoted run.
      return [decodeEscapes(trimmed.slice(0, -1))];
    }
    return [trimmed];
  }

  const parts: string[] = [];
  let index = 0;
  while (index < trimmed.length) {
    const character = trimmed[index] ?? "";
    if (/\s/u.test(character)) {
      index++;
      continue;
    }
    if (character === '"') {
      const run = scanQuotedRun(trimmed, index + 1);
      index = run.end;
      if (run.closed || run.value) parts.push(run.value);
      continue;
    }
    const run = scanBareRun(trimmed, index, false);
    index = run.end;
    if (run.value) parts.push(run.value);
  }
  return parts;
}

/**
 * Parse presentation-format content into whitespace separated tokens, each of
 * which is an independent `<character-string>`.
 *
 * Unlike {@link parseCharacterStrings}, bare runs are split on whitespace.
 * Use this for records whose RDATA is a fixed list of character-strings, such
 * as HINFO (`"CPU" "OS"`).
 */
export function parseCharacterStringTokens(raw: string | undefined): string[] {
  const trimmed = (raw ?? "").trim();
  if (!trimmed) return [];

  const tokens: string[] = [];
  let index = 0;
  while (index < trimmed.length) {
    const character = trimmed[index] ?? "";
    if (/\s/u.test(character)) {
      index++;
      continue;
    }
    if (character === '"') {
      const run = scanQuotedRun(trimmed, index + 1);
      index = run.end;
      if (run.closed || run.value) tokens.push(run.value);
      continue;
    }
    const run = scanBareRun(trimmed, index, true);
    index = run.end;
    if (run.value) tokens.push(run.value);
  }
  return tokens;
}

/**
 * The logical value of presentation-format content: adjacent character-strings
 * concatenate without a separator, exactly like a resolver joins TXT strings.
 *
 * Semantic parsers (SPF, DMARC, DKIM, …) should call this before inspecting
 * the payload so quoted and unquoted input behave identically.
 */
export function unquoteCharacterString(raw: string | undefined): string {
  return parseCharacterStrings(raw).join("");
}

/**
 * Escape the characters that may not appear literally inside a quoted
 * `<character-string>`: `"` and `\` are backslash escaped and control
 * characters use the `\DDD` decimal form.
 */
export function escapeCharacterString(value: string | undefined): string {
  let escaped = "";
  for (const character of value ?? "") {
    const codePoint = character.codePointAt(0) ?? 0;
    if (character === "\\") escaped += "\\\\";
    else if (character === '"') escaped += '\\"';
    else if (isControlCodePoint(codePoint))
      escaped += `\\${codePoint.toString(10).padStart(3, "0")}`;
    else escaped += character;
  }
  return escaped;
}

/** Wrap a single logical value in quotes, escaping its contents. */
export function quoteCharacterString(value: string | undefined): string {
  return `"${escapeCharacterString(value)}"`;
}

/**
 * Split a logical value into chunks of at most `maxBytes` UTF-8 bytes without
 * ever splitting a code point. Returns an empty array for an empty value.
 */
export function splitCharacterString(
  value: string | undefined,
  maxBytes: number = CHARACTER_STRING_MAX_BYTES,
): string[] {
  const input = value ?? "";
  if (!input) return [];
  const limit = Math.max(1, Math.floor(maxBytes));

  const chunks: string[] = [];
  let current = "";
  let size = 0;
  for (const character of input) {
    const characterSize = utf8Size(character.codePointAt(0) ?? 0);
    if (current && size + characterSize > limit) {
      chunks.push(current);
      current = "";
      size = 0;
    }
    current += character;
    size += characterSize;
  }
  if (current) chunks.push(current);
  return chunks;
}

/**
 * Serialize presentation-format content into a thoroughly valid quoted form.
 *
 * Quotes are balanced (a missing opening or closing quote is added), inner
 * quotes and backslashes are escaped, significant whitespace is preserved, and
 * payloads longer than 255 bytes become adjacent quoted character-strings.
 * Empty content normalizes to `""`.
 *
 * The result is stable: normalizing an already normalized value is a no-op.
 */
export function normalizeCharacterString(
  raw: string | undefined,
  options?: { maxBytes?: number },
): string {
  const value = unquoteCharacterString(raw);
  const chunks = splitCharacterString(
    value,
    options?.maxBytes ?? CHARACTER_STRING_MAX_BYTES,
  );
  if (!chunks.length) return '""';
  return chunks.map(quoteCharacterString).join(" ");
}

/** True when `raw` already equals its normalized serialization. */
export function isNormalizedCharacterString(
  raw: string | undefined,
  options?: { maxBytes?: number },
): boolean {
  const trimmed = (raw ?? "").trim();
  if (!trimmed) return false;
  return trimmed === normalizeCharacterString(trimmed, options);
}

File: src-tauri/crates/bc-dns-tools/src/character_string.rs
//! DNS `<character-string>` presentation format (RFC 1035 §3.3, §5.1).
//!
//! Record types whose RDATA is built from character-strings (TXT and the
//! TXT-shaped helpers SPF / DKIM / DMARC) are written in *presentation
//! format*: the payload may be bare, wrapped in double quotes, or expressed as
//! several adjacent quoted strings that concatenate into one logical value.
//!
//! This is the Rust mirror of `src/lib/dns/character-string.ts`. Semantic
//! parsers must call [`unquote_character_string`] before inspecting a payload,
//! so that quoted and unquoted input behave identically — otherwise a zone
//! that publishes `"v=spf1 mx -all"` reads as having no SPF record at all.

/// Decode a single `\X` / `\DDD` escape starting at `index`.
///
/// Returns the decoded text and the index just past the escape.
fn decode_escape_at(input: &[char], index: usize) -> (String, usize) {
    if index + 4 <= input.len() {
        let digits: String = input[index + 1..index + 4].iter().collect();
        if digits.len() == 3 && digits.chars().all(|c| c.is_ascii_digit()) {
            if let Ok(octet) = digits.parse::<u32>() {
                if octet <= 255 {
                    if let Some(decoded) = char::from_u32(octet) {
                        return (decoded.to_string(), index + 4);
                    }
                }
            }
        }
    }
    match input.get(index + 1) {
        None => ("\\".to_string(), index + 1),
        Some(next) => (next.to_string(), index + 2),
    }
}

/// Decode every escape sequence in `value`.
fn decode_escapes(value: &[char]) -> String {
    if !value.contains(&'\\') {
        return value.iter().collect();
    }
    let mut decoded = String::new();
    let mut index = 0;
    while index < value.len() {
        if value[index] == '\\' {
            let (text, next) = decode_escape_at(value, index);
            decoded.push_str(&text);
            index = next;
            continue;
        }
        decoded.push(value[index]);
        index += 1;
    }
    decoded
}

/// Count unescaped quotes and remember the position of the last one.
fn scan_quotes(value: &[char]) -> (usize, Option<usize>) {
    let mut count = 0usize;
    let mut last_index = None;
    let mut index = 0;
    while index < value.len() {
        if value[index] == '\\' {
            index += 2;
            continue;
        }
        if value[index] == '"' {
            count += 1;
            last_index = Some(index);
        }
        index += 1;
    }
    (count, last_index)
}

/// Whether `value` uses the quoted presentation form, including the damaged
/// shapes with only a leading or only a trailing quote.
fn is_quoted_form(value: &[char]) -> bool {
    if value.is_empty() {
        return false;
    }
    if value[0] == '"' {
        return true;
    }
    let (count, last_index) = scan_quotes(value);
    count % 2 == 1 && last_index == Some(value.len() - 1)
}

/// Read one quoted run starting just after its opening quote.
fn scan_quoted_run(input: &[char], start: usize) -> (String, usize, bool) {
    let mut value = String::new();
    let mut index = start;
    while index < input.len() {
        let character = input[index];
        if character == '\\' {
            let (text, next) = decode_escape_at(input, index);
            value.push_str(&text);
            index = next;
            continue;
        }
        if character == '"' {
            return (value, index + 1, true);
        }
        value.push(character);
        index += 1;
    }
    // Unmatched opening quote: repair by closing the string at end of input.
    (value, index, false)
}

/// Read one bare (unquoted) run up to the next unescaped quote.
fn scan_bare_run(input: &[char], start: usize) -> (String, usize) {
    let mut raw: Vec<char> = Vec::new();
    let mut index = start;
    while index < input.len() {
        let character = input[index];
        if character == '"' {
            break;
        }
        if character == '\\' {
            raw.extend_from_slice(&input[index..input.len().min(index + 2)]);
            index += 2;
            continue;
        }
        raw.push(character);
        index += 1;
    }
    (decode_escapes(&raw).trim().to_string(), index)
}

/// Parse presentation-format content into its logical character-strings.
///
/// Accepted shapes, all decoded to the same logical value:
/// - bare: `v=spf1 mx -all` → one string, kept verbatim
/// - quoted: `"v=spf1 mx -all"`
/// - adjacent strings: `"v=spf1 " "mx -all"` → two strings
/// - unmatched leading or trailing quote → repaired rather than rejected
pub fn parse_character_strings(raw: &str) -> Vec<String> {
    let trimmed: Vec<char> = raw.trim().chars().collect();
    if trimmed.is_empty() {
        return Vec::new();
    }

    if trimmed[0] != '"' {
        // Bare content is a single character-string: whitespace does not split it.
        if is_quoted_form(&trimmed) {
            // Unmatched trailing quote: repair by treating it as a quoted run.
            return vec![decode_escapes(&trimmed[..trimmed.len() - 1])];
        }
        return vec![trimmed.iter().collect()];
    }

    let mut parts = Vec::new();
    let mut index = 0;
    while index < trimmed.len() {
        let character = trimmed[index];
        if character.is_whitespace() {
            index += 1;
            continue;
        }
        if character == '"' {
            let (value, next, closed) = scan_quoted_run(&trimmed, index + 1);
            index = next;
            if closed || !value.is_empty() {
                parts.push(value);
            }
            continue;
        }
        let (value, next) = scan_bare_run(&trimmed, index);
        index = next;
        if !value.is_empty() {
            parts.push(value);
        }
    }
    parts
}

/// The logical value of presentation-format content: adjacent character-strings
/// concatenate without a separator, exactly like a resolver joins TXT strings.
///
/// Semantic parsers (SPF, DMARC, DKIM, …) should call this before inspecting
/// the payload so quoted and unquoted input behave identically.
pub fn unquote_character_string(raw: &str) -> String {
    parse_character_strings(raw).join("")
}

/// Whether a TXT payload is an SPF record.
///
/// The single SPF detector on the Rust side, mirroring `isSpfRecord` in
/// `src/lib/dns/spf.ts`. It lives here rather than in `bc-spf` because it needs
/// [`unquote_character_string`], and `bc-dns-tools` already depends on
/// `bc-spf` — the reverse edge would be a dependency cycle.
///
/// RFC 7208 §4.5: the version section is `v=spf1` followed by whitespace or the
/// end of the record, so `v=spf1foo` is a different record that merely shares a
/// prefix. Content may be bare, quoted, or split into adjacent
/// character-strings; all forms answer identically.
pub fn is_spf_record(content: &str) -> bool {
    let value = unquote_character_string(content);
    let value = value.trim();
    let Some(rest) = value.get(..6) else {
        return false;
    };
    if !rest.eq_ignore_ascii_case("v=spf1") {
        return false;
    }
    match value[6..].chars().next() {
        None => true,
        Some(next) => next.is_whitespace(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bare_content_is_returned_verbatim() {
        assert_eq!(unquote_character_string("v=spf1 mx -all"), "v=spf1 mx -all");
    }

    #[test]
    fn quoted_content_loses_its_quotes() {
        assert_eq!(
            unquote_character_string("\"v=spf1 mx -all\""),
            "v=spf1 mx -all"
        );
        assert_eq!(
            unquote_character_string("   \"v=spf1 mx -all\"  "),
            "v=spf1 mx -all"
        );
    }

    #[test]
    fn adjacent_character_strings_concatenate_without_a_separator() {
        assert_eq!(
            unquote_character_string("\"v=spf1 \" \"mx -all\""),
            "v=spf1 mx -all"
        );
        assert_eq!(parse_character_strings("\"a\" \"b\""), vec!["a", "b"]);
    }

    #[test]
    fn unmatched_quotes_are_repaired_rather_than_rejected() {
        assert_eq!(
            unquote_character_string("\"v=spf1 mx -all"),
            "v=spf1 mx -all"
        );
        assert_eq!(
            unquote_character_string("v=spf1 mx -all\""),
            "v=spf1 mx -all"
        );
    }

    #[test]
    fn escapes_inside_a_quoted_run_are_decoded_not_treated_as_delimiters() {
        assert_eq!(unquote_character_string("\"a\\\"b\""), "a\"b");
        assert_eq!(unquote_character_string("\"a\\\\b\""), "a\\b");
        assert_eq!(unquote_character_string("\"a\\032b\""), "a b");
    }

    #[test]
    fn whitespace_inside_quotes_is_preserved() {
        assert_eq!(unquote_character_string("\"a  b\""), "a  b");
    }

    #[test]
    fn empty_content_yields_no_strings() {
        assert!(parse_character_strings("").is_empty());
        assert!(parse_character_strings("   ").is_empty());
    }

    #[test]
    fn spf_is_detected_in_every_presentation_shape() {
        for content in [
            "v=spf1 mx -all",
            "\"v=spf1 mx -all\"",
            "  \"v=spf1 mx -all\"  ",
            "\"v=spf1 \" \"mx -all\"",
            "\"V=SPF1 MX -ALL\"",
            "v=spf1",
            "\"v=spf1\"",
        ] {
            assert!(
                is_spf_record(content),
                "expected SPF detection for {content}"
            );
        }
    }

    #[test]
    fn the_spf_version_token_needs_a_boundary() {
        for content in ["v=spf1foo bar", "v=spf2 mx -all", "", "\"\"", "v=spf"] {
            assert!(
                !is_spf_record(content),
                "expected no SPF match for {content}"
            );
        }
    }
}
